linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -698,45 +695,97 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
+    val (restParamLocalDef, newRestParam) = {
+      // Option#unzip
+      restParam.map(newParamReplacement(_)) match {
+        case None         => (None, None)
+        case Some((x, y)) => (Some(x), Some(y))
+      }
     }
 
-    withCaptures(newCaptureParamsWithUse, tcaptureValues) { (finalCaptureParams, finalCaptureValues, cont1) =>
-      val newClosure = Closure(arrow, finalCaptureParams, newParams, newRestParam, newBody, finalCaptureValues)
-      val newTree = PreTransTree(newClosure, RefinedType(AnyType, isExact = false, isNullable = false))
-      cont1(newTree)
+    val thisLocalDef =
+      if (arrow) None
+      else Some(newThisReplacement(AnyType))
+
+    val innerEnv = OptEnv.Empty
+      .withThisLocalDef(thisLocalDef)
+      .withLocalDefs(paramLocalDefs)
+      .withLocalDefs(restParamLocalDef.toList)
+
+    transformCapturingBody(captureParams, tcaptureValues, body, innerEnv) {
+        (newCaptureParams, newCaptureValues, newBody) =>
+      val newClosure = Closure(arrow, newCaptureParams, newParams, newRestParam, newBody, newCaptureValues)
+      PreTransTree(newClosure, RefinedType(AnyType, isExact = false, isNullable = false))
     } (cont)
   }
 
-  private def withCaptures(newCaptureParamsWithUse: List[(ParamDef, IsUsed)],
-      tcaptureValues: List[PreTransform])(
-      buildInner: (List[ParamDef], List[Tree], PreTransCont) => TailRec[Tree])(
-      cont: PreTransCont)(implicit scope: Scope, pos: Position) = {
-    val bindings = {
-      newCaptureParamsWithUse.iterator.zip(tcaptureValues.iterator).map { case ((param, _), value) =>
-        Binding.temp(param.name.name, param.ptpe, mutable = false, value)
-      }.toList
+  private def transformCapturingBody(captureParams: List[ParamDef],
+      tcaptureValues: List[PreTransform], body: Tree, innerEnv: OptEnv)(
+      inner: (List[ParamDef], List[Tree], Tree) => PreTransTree)(cont: PreTransCont)(
+      implicit scope: Scope, pos: Position): TailRec[Tree] = {
+    /* Process captures.
+     *
+     * This is different than normal param replacement:
+     *
+     * - We inline literals into the closure.
+     * - We introduce bindings for the (remaining) capture values. This allows to
+     *   eliminate captures that are not used (but preserve side effects).
+     * - If the capture value is a VarRef, we give the capture param the exact
+     *   same name. This is to help the FunctionEmitter eliminate IIFEs.
+     */
+    val captureParamLocalDefs = List.newBuilder[(LocalName, LocalDef)]
+    val newCaptureParamDefsAndRepls = List.newBuilder[(ParamDef, ReplaceWithVarRef)]
+    val captureValueBindings = List.newBuilder[Binding]
+
+    for ((paramDef, tcaptureValue) <- captureParams.zip(tcaptureValues)) {
+      val ParamDef(ident @ LocalIdent(paramName), originalName, ptpe, mutable) = paramDef
+
+      assert(!mutable, s"Found mutable capture at ${paramDef.pos}")
+
+      def withCaptureParam(newName: LocalName) = {
```

Comment:
I keep misreading this as returning a value, but it's actually side-effecting. Perhaps rename it to `addCaptureParam`, and/or write the explicit `: Unit` result type?